File: src/Extensions/Components/BootstrapBlazor.Live2DDisplay/Components/Live2DDisplay/Live2DDisplay.razor.cs
using Microsoft.AspNetCore.Components;

namespace BootstrapBlazor.Components;

/// <summary>
/// Live2DDisplay Live2D模型组件
/// </summary>
public partial class Live2DDisplay
{
    private bool _isInit;

    private string? _source;
    private double _scale;
    private int _xoffset;
    private int _yoffset;
    private bool _isDraggble;
    private bool _addHitAreaFrames;
    private string? _backgroundColor;
    private bool _backgroundAlpha;

    private LivePosition _position;

    private string? ClassString { get; set; }

    /// <summary>
    /// 获得/设置 ModelSource 模型源
    /// </summary>
    [Parameter]
    public string? Source { get; set; }

    /// <summary>
    /// 获得/设置 Scale 模型比例
    /// </summary>
    [Parameter]
    public double Scale { get; set; }

    /// <summary>
    /// 获得/设置 Xoffset 模型X轴偏移
    /// </summary>
    [Parameter]
    public int Xoffset { get; set; }

    /// <summary>
    /// 获得/设置 Yoffset 模型Y轴偏移
    /// </summary>
    [Parameter]
    public int Yoffset { get; set; }

    /// <summary>
    /// 获得/设置 IsDraggble 是否可以拖动模型 unrealized
    /// </summary>
    //[Parameter]
    private bool IsDraggble { get; set; } = false;

    /// <summary>
    /// 获得/设置 HitAreaFrames 是否渲染鼠标命中区域框
    /// </summary>
    [Parameter]
    public bool AddHitAreaFrames { get; set; }

    /// <summary>
    /// 获得/设置 Position 模型默认显示位置
    /// </summary>
    [Parameter]
    public LivePosition Position { get; set; }

    /// <summary>
    /// 获得/设置 BackgroundColor 模型画板背景颜色
    /// </summary>
    [Parameter]
    public string BackgroundColor { get; set; } = "#000000";

    /// <summary>
    /// 获得/设置 BackgroundAlpha 模型画板是否透明
    /// </summary>
    [Parameter]
    public bool BackgroundAlpha { get; set; }

    /// <inheritdoc/>
    protected override void OnInitialized()
    {
        base.OnInitialized();
        ClassString = CssBuilder.Default("bb-model-default")
            .AddClassFromAttributes(AdditionalAttributes)
            .Build();
    }

    /// <inheritdoc/>
    protected override async Task OnParametersSetAsync()
    {
        if (_isInit)
        {
            if (Source != _source)
            {
                await InvokeVoidAsync("changeSource", new { Id, Source, Scale, X = Xoffset, Y = Yoffset, IsDraggble, AddHitAreaFrames });
                _source = Source;
            }
            if (Scale != _scale)
            {
                await InvokeVoidAsync("changeScale", Id, Scale);
                _scale = Scale;
            }
            if (Xoffset != _xoffset || Yoffset != _yoffset)
            {
                await InvokeVoidAsync("changeXY", Id, Xoffset, Yoffset);
                _xoffset = Xoffset;
                _yoffset = Yoffset;
            }
            if (!(IsDraggble && _isDraggble))
            {
                await InvokeVoidAsync("changeDraggble", Id, IsDraggble);
                _isDraggble = IsDraggble;
            }
            if (!(AddHitAreaFrames && _addHitAreaFrames))
            {
                await InvokeVoidAsync("addHitAreaFrames", Id, AddHitAreaFrames);
                _isDraggble = IsDraggble;
            }
            if (Position != _position)
            {
                ClassString = $"bb-model-{Position}";
                _position = Position;
            }
            if ((BackgroundColor != _backgroundColor) || (BackgroundAlpha != _backgroundAlpha))
            {
                await InvokeVoidAsync("changebackground", Id, BackgroundAlpha, BackgroundColor);
                _backgroundColor = BackgroundColor;
                _backgroundAlpha = BackgroundAlpha;
            }
        }
    }

    /// <inheritdoc/>
    protected override async Task InvokeInitAsync()
    {
        await InvokeVoidAsync("init", new { Id, Source, Scale, X = Xoffset, Y = Yoffset, IsDraggble, AddHitAreaFrames, BackgroundAlpha, BackgroundColor });
        _isInit = true;
        _source = Source;
        _scale = Scale;
        _xoffset = Xoffset;
        _yoffset = Yoffset;
        _position = Position;
        _backgroundColor = BackgroundColor;
        _backgroundAlpha = BackgroundAlpha;
        _addHitAreaFrames = AddHitAreaFrames;
        _isDraggble = IsDraggble;
    }
}

File: src/Extensions/Components/BootstrapBlazor.Live2DDisplay/Components/Live2DDisplay/Live2DDisplay.razor.js
import { addLink, addScript } from '../../../BootstrapBlazor/modules/utility.js'
import Data from '../../../BootstrapBlazor/modules/data.js'
import EventHandler from '../../../BootstrapBlazor/modules/event-handler.js'

export async function init(op) {
    await addLink("./_content/BootstrapBlazor.Live2DDisplay/BootstrapBlazor.Live2DDisplay.bundle.scp.css")
    await addScript("./_content/BootstrapBlazor.Live2DDisplay/live2dcubismcore.min.js")
    await addScript("./_content/BootstrapBlazor.Live2DDisplay/live2d.min.js")
    await addScript("./_content/BootstrapBlazor.Live2DDisplay/pixi.min.js")
    await addScript("./_content/BootstrapBlazor.Live2DDisplay/index.min.js")
    await addScript("./_content/BootstrapBlazor.Live2DDisplay/extra.min.js")

    const model = await createModel(op);
    if (model) {
        const el = document.getElementById(op.id);
        el.style.zIndex = '9999';
        el.style.opacity = '1';

        const canvas = document.createElement('canvas');

        el.appendChild(canvas);

        const app = new PIXI.Application({
            view: canvas,
            autoStart: true
        });
        app.stage.addChild(model);

        const data = { app, model, el, canvas, op };
        Data.set(op.id, data);

        resizeTo(data);
        changeDraggable(op.id, op.isDraggable);
        changeBackground(op.id, op.backgroundAlpha, op.backgroundColor);
    }
}


function createHitAreaFrames(op, model) {
    model.children = [];
    if (op.addHitAreaFrames) {
        // handle tapping
        model.on("hit", (hitAreas) => {
            if (hitAreas.includes("Body")) {
                model.motion("Tap");
                model.motion("tap_body");
            }

            if (hitAreas.includes("Head")) {
                model.expression();
            }
        });
        const hitAreaFrames = new PIXI.live2d.HitAreaFrames();
        model.addChild(hitAreaFrames);
        hitAreaFrames.visible = op.addHitAreaFrames;
    }
}

function resizeTo(data) {
    if (data.op.isDraggble) {
        data.app.resizeTo = window;
    }
    else {
        var h = Math.ceil(data.model.height);
        var w = Math.ceil(data.model.width);

        data.el.style.width = w + 'px';
        data.el.style.height = h + 'px';

        data.canvas.style.width = data.el.style.width;
        data.canvas.style.height = data.el.style.height;

        data.canvas.width = w;
        data.canvas.height = h;

        data.app.resizeTo = data.el;
    }
}

export async function changeSource(op) {
    const data = Data.get(op.id);
    data.op = op;
    data.app.stage.removeChild(data.model);
    data.model = await createModel(op);
    if (data.model) {
        data.app.stage.addChild(data.model);
        data.model.scale.set(data.op.scale);
        resizeTo(data);
    }
}

export function changeScale(id, scale) {
    const data = Data.get(id);
    data.op.scale = scale;
    data.model.scale.set(data.op.scale);
    resizeTo(data);
}

export function changeXY(id, x, y) {
    const data = Data.get(id);
    data.op.x = x;
    data.op.y = y;
    data.model.x = x;
    data.model.y = y;
    resizeTo(data);
}

export function changeBackground(id, backgroundAlpha, backgroundColor) {
    if (backgroundColor) {
        if (backgroundColor.startsWith('#')) {
            const data = Data.get(id);
            backgroundColor = backgroundColor.replace("#", "0x");
            data.op.backgroundAlpha = backgroundAlpha;
            data.op.backgroundColor = backgroundColor;
            data.app.renderer.backgroundColor = backgroundColor;
            data.app.renderer.backgroundAlpha = backgroundAlpha;
        }
    }
}

export function changeDraggable(id, draggble) {
    const data = Data.get(id);
    data.op.isDraggble = draggble;

    // 获取要拖动的元素
    var el = data.el;

    if (!draggble) {
        EventHandler.off(el, 'mousedown');
        EventHandler.off(document, 'mousemove');
        EventHandler.off(document, 'mouseup');

        el.style.cursor = 'none';
        el.style.pointerEvents = 'none';
    } else {
        el.style.cursor = 'move';
        el.style.pointerEvents = 'auto';
        // 初始化拖动的起始位置
        var offsetX = 0;
        var offsetY = 0;

        // 当鼠标按下时触发
        EventHandler.on(el, 'mousedown', (e) => {
            // 计算鼠标在元素内的相对位置
            offsetX = e.clientX - el.offsetLeft;
            offsetY = e.clientY - el.offsetTop;

            // 添加鼠标移动和释放事件监听器
            // 当鼠标移动时触发
            EventHandler.on(document, 'mousemove', (e) => {
                // 阻止默认的拖动行为
                e.preventDefault();

                // 计算元素的新位置
                var x = e.clientX - offsetX;
                var y = e.clientY - offsetY;

                // 设置元素的新位置
                el.style.left = x + "px";
                el.style.top = y + "px";
            });

            // 当鼠标释放时触发
            EventHandler.on(document, 'mouseup', () => {
                // 移除鼠标移动和释放事件监听器
                EventHandler.off(document, 'mousemove');
                EventHandler.off(document, 'mouseup');
            });
        })
    }
}

export function addHitAreaFrames(id, isaddHitAreaFrames) {
    const data = Data.get(id);
    data.op.addHitAreaFrames = isaddHitAreaFrames;
    createHitAreaFrames(data.op, data.model);
}

async function createModel(op) {
    const model = await PIXI.live2d.Live2DModel.from(op.source);
    model.scale.set(op.scale);
    model.x = op.x;
    model.y = op.y;
    model.dragging = op.isDraggble;
    if (op.addHitAreaFrames) {
        createHitAreaFrames(op, model);
    }
    return model;
}
